src/Data/CAS/RocksDB.hs

```diff
@@ -757,3 +770,79 @@ decIterKey it k = case B.splitAt (B.length namespace) k of
     namespace = _rocksDbTableIterNamespace it
 {-# INLINE decIterKey #-}
 
+data Checkpoint
+
+foreign import ccall unsafe "rocksdb\\c.h rocksdb_checkpoint_object_create" 
+  rocksdb_checkpoint_object_create :: C.RocksDBPtr -> Ptr CString -> IO (Ptr Checkpoint)
+
+foreign import ccall unsafe "rocksdb\\c.h rocksdb_checkpoint_create"
+  rocksdb_checkpoint_create :: Ptr Checkpoint -> CString -> CULong -> Ptr CString -> IO ()
+
+foreign import ccall unsafe "rocksdb\\c.h rocksdb_checkpoint_object_destroy"
+  rocksdb_checkpoint_object_destroy :: Ptr Checkpoint -> IO ()
+
+checked :: String -> Ptr CString -> IO a -> IO a
+checked whatWasIDoing errPtr act = do
+  r <- act
+  err <- peek errPtr
+  unless (err == nullPtr) $ do
+    errStr <- B.packCString err
+    let msg = unwords ["error while", whatWasIDoing <> ":", B8.unpack errStr]
+    free err
+    fail msg
+  return r
+
+-- to unconditionally flush the WAL log before making the checkpoint, set logSizeFlushThreshold to zero. 
+-- to *never* flush the WAL log, set logSizeFlushThreshold to maxBound :: CULong.
+checkpointRocksDb :: RocksDb -> CULong -> FilePath -> IO ()
+checkpointRocksDb RocksDb { _rocksDbHandle = R.DB dbPtr _ } logSizeFlushThreshold path = 
+    alloca (\errPtr -> do
+      poke errPtr (nullPtr :: CString)
+      let 
+          mkCheckpointObject = 
+            checked "creating checkpoint object" errPtr $ 
+              rocksdb_checkpoint_object_create dbPtr errPtr
+          mkCheckpoint cp =
+            withCString path (\path' -> 
+              checked "creating checkpoint" errPtr $ 
+                rocksdb_checkpoint_create cp path' logSizeFlushThreshold errPtr
+              )
+      bracket mkCheckpointObject rocksdb_checkpoint_object_destroy mkCheckpoint 
+    )
+
+foreign import ccall unsafe "rocksdb\\c.h rocksdb_approximate_sizes" 
+    rocksdb_approximate_sizes
+        :: C.RocksDBPtr 
+        -> {- input: number of key ranges -} CInt 
+        -> {- input: array of range start keys -} Ptr CString 
+        -> {- input: array of range start key lengths -} Ptr CSize 
+        -> {- input: array of range end keys -} Ptr CString 
+        -> {- input: array of range end key lengths -} Ptr CSize 
+        -> {- output: array of sizes -} Ptr CULong 
+        -> {- errptr -} Ptr CString 
+        -> IO ()
+
+approxTableSizeRocksDb :: RocksDb -> RocksDbTable k v -> IO (Maybe CULong)
+approxTableSizeRocksDb RocksDb { _rocksDbHandle = R.DB dbPtr _ } table = do
+    bounds <- withTableIter table $ \iter -> 
+        (,)
+            <$> (tableIterFirst iter *> R.iterKey (_rocksDbTableIter iter)) 
+            <*> (tableIterLast iter *> R.iterKey (_rocksDbTableIter iter))
+    forM (sequenceOf each bounds) $ \(minKey, maxKey) ->
```

Comment:
Isn't `bounds` just a single tuple? So, why treat it as a list. What am I missing?

Reply:
It's a pair of `Maybe`s, this makes it a `Maybe` of a pair.

Reply:
It's `Maybe` of a pair, this makes it a pair of `Maybe`.

Reply:
Ah, OK. Have to refresh my lens skills :-)